intf/intf.go

```diff
@@ -0,0 +1,156 @@
+	// ARPAdd indicates that the ARP neighbour was added.
+	ARPAdd
+	// ARPDelete inidcates that the ARP neighbour was deleted.
+	ARPDelete
+)
+
+// ARPUpdate is used to describe a change to the ARP table.
+type ARPUpdate struct {
+	// Type indicates whether the event was an add or delete.
+	Type ARPEvent
+	// Neigh is the neighbour that the change corresponds to
+	Neigh ARPEntry
+}
+
+// NetworkAccessor is an interface implemented by the underlying system access
+// interface. Through implementing the NetworkAccessor interface it is possible
+// to use the functions within this package whilst using a different underlying
+// implementation (e.g., netlink on Linux to access the kernel).
+type NetworkAccessor interface {
+	// Links returns a set of links that are present on the local system.
+	Interfaces() ([]*Interface, error)
+	// Link retrieves the link with the specified name.
+	Interface(name string) (*Interface, error)
+	// AddressList lists the IP addresses configured on a particular interface.
+	InterfaceAddresses(name string) ([]*net.IPNet, error)
+	// AddressAdd adds address ip to the interface name.
+	AddInterfaceIP(name string, ip *net.IPNet) error
+	// ARPList lists the set of ARP neighbours on the system.
+	ARPList() ([]*ARPEntry, error)
+	// ARPSubscribe writes changes to the ARP table to the channel updates, and uses
+	// done to indicate that the subscription is complete.
+	ARPSubscribe(updates chan ARPUpdate, done chan struct{}) error
+}
+
+// unimplementedAccessor is an accessor interface that returns unimplemented
+// for all underlying methods.
+type unimplementedAccessor struct{}
+
+// Interface returns unimplemented for the interface call.
+func (u unimplementedAccessor) Interface(_ string) (*Interface, error) {
+	return nil, fmt.Errorf("unimplemented")
+}
+
+// Interfaces returns unimplemented when asked to list interfaces.
+func (u unimplementedAccessor) Interfaces() ([]*Interface, error) {
+	return nil, fmt.Errorf("unimplemented")
+}
+
+// InterfaceAddresses returns unimplemented when asked for list IP addresses.
+func (u unimplementedAccessor) InterfaceAddresses(_ string) ([]*net.IPNet, error) {
+	return nil, fmt.Errorf("unimplemented")
+}
+
+// AddInterfaceIP returns unimplemented when asked to add an interface.
+func (u unimplementedAccessor) AddInterfaceIP(_ string, _ *net.IPNet) error {
+	return fmt.Errorf("unimplemented")
+}
+
+// ARPList returns unimplemented when asked to list ARP entries.
+func (u unimplementedAccessor) ARPList() ([]*ARPEntry, error) {
+	return nil, fmt.Errorf("unimplemented")
+}
+
+// ARPSubscribe returns unimplemented when asked to subscribe to ARP changes.
+func (u unimplementedAccessor) ARPSubscribe(_ chan ARPUpdate, _ chan struct{}) error {
+	return fmt.Errorf("unimplemented")
+}
+
+// accessor is the implementation of the network accessor that should be used. It
+// should be set by init() [which may be a platform-specific initiation].
+var accessor NetworkAccessor
```

Comment:
I see that it is the init() method in accessor_linux.go that initializes this. what gates whether or not init() is called for specific platforms? Is the expected use that the user calls `accessor_<their_platform>.init()` to run their platform-specific initialization?